File: src/main/java/org/engine/GameLoop.java
package org.engine;

import org.graphics.Render;
import org.world.World;

public class GameLoop {
    long last_time = System.nanoTime();
    private boolean running = true;
    private int currentFPS = 0;

    public void start() {
        Thread game = new Thread(() -> {
            while (running) {
                long time = System.nanoTime();
                float delta_time = (time - last_time) / 1000000000f;
                last_time = time;
                World.update(delta_time);
                Render.render();
                if (delta_time == 0) {
                    currentFPS = 0;
                } else {
                    currentFPS = (int) (1 / delta_time);
                }
            }
            AudioManager.cleanup();
            DiscordHandler.shutdown();
        });
        game.setName("GameLoop");
        game.start();
    }

    public void setRunning(boolean running) {
        this.running = running;
    }

    public int getCurrentFPS() {
        return currentFPS;
    }
}

File: src/main/java/org/engine/GameLoop.kt
package org.engine

import org.engine.AudioManager.cleanup
import org.engine.DiscordHandler.shutdown
import org.engine.DiscordHandler.updatePresence
import org.graphics.Render
import org.world.World.update

class GameLoop {
    private var lastTime = System.nanoTime()
    private var running = true
    var currentFPS = 0
        private set

    fun start() {
        val game = Thread {
            while (running) {
                val time = System.nanoTime()
                val deltaTime = (time - lastTime) / 1000000000f
                lastTime = time
                update(deltaTime)
                updatePresence()
                Render.render()
                currentFPS = if (deltaTime == 0f) {
                    0
                } else {
                    (1 / deltaTime).toInt()
                }
            }
            cleanup()
            shutdown()
        }
        game.name = "GameLoop"
        game.start()
    }

    fun setRunning(running: Boolean) {
        this.running = running
    }
}